### DISCUSSION TITLE
KOReader Position Sync API Compatibility

### DISCUSSION THREAD

**Discussion by alexcash** [Kareadita/Kavita#2590] [Ideas]
### Idea Description

KOReader has a (default provided) plugin that implements reading position sync, given a compliant server. Source linked below:

https://github.com/koreader/koreader/tree/master/plugins/kosync.koplugin

Seems to be fairly simple in design: GET and PUT a percentage completed for a given document. It allows identification based on document checksum or filename.

I use OPDS-PS for manga while I'm on my home network, but for books I do full downloads (also via OPDS). That way I can turn on airplane mode and save battery, take my ereader out of the house, etc. But that means if I want to swap over to my computer, phone, or iPad I don't really have a way to pick up where I left off. If Kavita provided a KOReader sync compliant endpoint, it would make swapping between devices easier.

### Duration of Using Kavita

2 Weeks

### Idea Category

Feature Enhancement

### Agreement

- [X] I agree that this is solely for submitting ideas, and I will search for existing ideas before posting.

**Comment by majora2007** [comment]
I believe Kavita is already supported with KOReader. 

See https://github.com/koreader/koreader/pull/9750

**Comment by alexcash** [comment]
It's possible I'm mistaken, but I think that only works if you're using OPDS-PS. Support for the progress-sync API would work for books that had been downloaded for offline reading via OPDS (non page streaming).

**Reply by majora2007** [reply]
I don't think that's the case. Opds-PS doesn't work with epubs as it's a simple image based. 

Would you mind testing? Last I spoke with KOReader, they informed there was syncing similar to tachiyomi.

**Reply by majora2007** [reply]
Linking this as this was my last discussion:
https://github.com/koreader/koreader/issues/11105

**Reply by alexcash** [reply]
I'm pretty sure. If you look at the patch for https://github.com/koreader/koreader/pull/9750 the methods are all namespaced under OPDS-PSE e.g. ` OPDSPSE:getLastPage` and as far as I can tell `getLastPage` is only called from inside `streamPages`. This mirrors my experience using the app.

![FileManager_2024-01-09_193110](https://github.com/Kareadita/Kavita/assets/1558955/663900c9-d652-4a2d-89b4-015f62e5529e)

Here is a screenshot of what happens when you tap on a book listed in the Kavita OPDS library from KOReader. If it were manga, tapping "Page stream" would load up the book from the last read page. Additionally, when I close that Manga the last read page is saved to Kavita. That works like a charm. But as you say, you can't page stream an epub. Tapping those options simply loads an image of the koreader logo. "Book cover" and "Book information" show details about the book. All the remaining buttons pertain to downloading the book in its entirety. Once the book is downloaded, it doesn't appear to be any different than any other book dragged into the "Books" folder of the device internal storage. As far as I can tell it's totally untethered from Kavita. My progress isn't synced to Kavita, and as far as I can tell the only UI related to my reading position is labeled "Progress Sync" and prompts me to log into a KOReader sync server -- which I believe is the API I linked in my initial post.

**Comment by dandersch** [comment]
> I don't think that's the case. Opds-PS doesn't work with epubs as it's a simple image based.

@majora2007  Is there a technical reason that makes page streaming support for epubs/pdfs infeasible? I saw that the OPDS-PS standard doesn't specify support for any other file types, but I don't see why a page streaming extension should remain exclusive to comic books.

**Reply by majora2007** [reply]
Yes. The spec was created and allows syncing on a page by page level. Epubs do not work on a page-by-page level, but rather are free form text, css, etc that are usually laid out in a manner to look like a page. But from your reader, pages 1-10 could actually be just the first page on that epub, or the whole chapter could just be one page. 

The spec only allows for page level. While Kavita can support OPDS-PS, no epub client does support this due to the complexity and how bad the end result would be for them. Epubs cannot be streamed they are in fact html files, css, etc. 

This is why there are no epubs that support syncing via OPDS-PS and I don't expect to see that to change. The best course of action is a dedicated API for these softwares that support OPDS (Calibre, Komga, Kavita, Stump). 

The official spec is here: https://vaemendis.net/opds-pse/

Hope this helps.

**Reply by dandersch** [reply]
Thanks for the explanation. Am I right to assume that if the server knew about the readers screen size, font size, margins etc., then it could potentially construct the page layout and serve a single page epub for every streamed page?

What about PDFs? Shouldn't PS support be more straightforward to implement because ebook PDFs are usually page based?

**Comment by majora2007** [comment]
A user was kind enough to implement this. We are wrapping up the implementation and performing testing with some community members. 
https://github.com/Kareadita/Kavita/pull/3311

Edit: There is some last minute polish however the submitter is busy. If anyone can code, we could use a few small things to get this ready for the masses.

**Reply by alexcash** [reply]
woah! huge shout out to all contributors!!! I'll deploy canary and give it a test.

**Reply by majora2007** [reply]
It's not on canary yet. The creator needs to handle a few situations. I will reach out once it's released. 

Please remember to never use canary with your main instance. It will break your DB.

**Reply by majora2007** [reply]
@alexcash Would you be able to provide me with some understanding of how this can be tested? I have 2 users with Koreader devices but we aren't sure how to actually setup for testing and validating.

**Reply by alexcash** [reply]
@majora2007 

- open a book in koreader
- tap the top third of the screen to open the menu
- tap the screwdriver/wrench icon
- tap "progress sync"
- tap custom sync server and setup the address
- you can then either manually push/pull sync progress or configure automatic settings.

**Reply by balgerion** [reply]
What server address should we put in sync server ? (Kavita or opds)

**Reply by scare376** [reply]
koreader progress sync support is currently on hold

**Reply by balgerion** [reply]
Is this still in plans ?  is there any good news :)

**Reply by martinjose** [reply]
Really hoping this is still in the works. 😬

**Reply by Tycho-S** [reply]
It looks like the related merge request is now closed: https://github.com/Kareadita/Kavita/pull/3311

I wonder if this means if it's going ahead or that it's officially been cancelled? I assumed "Closed" meant cancelled, but on the other hand a bot posted this: "moved this from In progress to Done in [ v0.8 - PDF & Comic Love ]"

Would it really be coming in 0.8? A progress sync capability for epub would be a gamechanger IMO.

**Reply by scare376** [reply]
That PR is very old so it was closed as there has been no response from the contributor. There are no current plans to to work on the implementation but will welcome it if the original contributor comes back or someone else wants to bring the PR up to date and correct the issues

**Reply by Tycho-S** [reply]
Ok thanks! I hope they will come back and finish it eventually. I don't have the skills for it otherwise I'd do it.

I moved to Kavita + Koreader and I love it, after being locked into a commercial platform for decades. The position sync is really the one thing I'm really missing because I read on so many devices. I had given up on it because the OPDS-PS is only page/image based and I only really read ePubs, but the progress-sync API should fix that if I understand correctly.

**Reply by majora2007** [reply]
While the koreader PR has stalled out, I'm still planning Kobo sync support. Still very open if anyone wants to contribute.

**Reply by majora2007** [reply]
The Koreader PR is in active testing as @tgkenney has picked it up.

**Comment by majora2007** [comment]
Koreader PR has been merged and will be in the v0.8.7 release. Thanks again @tgkenney for finishing it off and @MFDeAngelo for the initial implementation.
